boruta/1-click-boruta/script.whizzml

```diff
@@ -0,0 +1,386 @@
+                        objective-ids))
+    (if (not (member? objective-name objective-names))
+      (raise {"message" (str "Failed to find the objective name in the dataset"
+                             " choosable fields.")
+              "code" 101})
+      objective-name)))
+
+
+;; boruta
+;;
+;; Marks fields as important, unimportant or neutral according to their
+;; importance compared to shadow fields (copy fields using random values)
+;;
+;; Inputs:
+;;   dataset-id: (dataset-id) ID of the origin dataset
+;;   objective: (string) Name of the objective field
+;;   max-runs: (integer) Maximum number of iterations
+;;   min-gain: (number) Increment to define the interval of minimum gain
+;;                      around the maximum importance of shadow fields
+;;
+;; Output: (map) Map of important fields and unimportant fields
+
+(define (boruta dataset-id objective max-runs min-gain)
+  (let (ds (fetch dataset-id)
+        ;; original fields are stored for later use
+        ori-fields (get ds "fields" {})
+        field-ids (keys ori-fields)
+        f-names (map (lambda (x) (get-in ori-fields [x "name"] false))
+                     field-ids)
+        init-rank (reduce (lambda (x y) (assoc x (get-in ori-fields
+                                                         [y "name"]
+                                                         false) 0))
+                          {}
+                          field-ids)
+        objective (if (= objective "")
+                      (get-in ori-fields
+                              [(dataset-get-objective-id dataset-id) "name"]
+                              false)
+                      (validate-objective-name objective ds)))
+    (loop (f-rank init-rank fields ori-fields ds-id dataset-id runs 0)
+      ;; Repeats the procedure until the importance is assigned for all
+      ;; the attributes, or the algorithm has reached the previously set
+      ;; limit of the random forest runs.
+      (if (or (= runs max-runs) (full f-rank))
+          ;; as runs go by, the fields structure is modified and unimportant
+          ;; fields are removed. Thus we need f-names to contain the original
+          ;; list of names to know which are unimportant
+          (group-importance f-rank fields f-names objective)
+          ;; Loop body: Extends the information system by adding copies of all
+          ;; variables (the information system is always extended by at least
+          ;; 5 shadow attributes, even if the number of attributes in the
+          ;; original set is lower than 5)
+          ;; Shuffle the added attributes to remove their correlations with
+          ;; the response.
+          (let (sh-ds-info (shadow-ds ds-id fields objective)
+                sh-ds-id (head sh-ds-info)
+                sh-ids (head (tail sh-ds-info))
+                fields (nth sh-ds-info 2)
+                to-names (lambda (x) (get-in fields [x "name"] false))
```

Comment:
i was confused by the name of this function, which sounds as if it returned a list of names... perhaps `to-name`?